modules/mongodb/mongodb.go

```diff
@@ -0,0 +1,114 @@
+package mongo
+
+import (
+	"context"
+	"time"
+
+	"github.com/netdata/go.d.plugin/agent/module"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func init() {
+	module.Register("mongodb", module.Creator{
+		Defaults: module.Defaults{
+			UpdateEvery:        module.UpdateEvery,
+			AutoDetectionRetry: module.AutoDetectionRetry,
+			Priority:           module.Priority,
+			Disabled:           false,
+		},
+		Create: func() module.Module { return New() },
+	})
+}
+
+func New() *Mongo {
+	return &Mongo{
+		Config: Config{
+			Timeout: defaultTimeout,
+			Uri:     defaultUri,
+		},
+		charts: &module.Charts{},
+	}
+}
+
+type Mongo struct {
+	module.Base
+	Config `yaml:",inline"`
+	client *mongo.Client
+	charts *module.Charts
+}
+
+func (m *Mongo) Init() bool {
+	m.Infof("initializing mongodb")
+	if m.Uri == "" {
+		m.Errorf("connection URI is empty")
+		return false
+	}
+
+	var err error
+	m.charts, err = m.initCharts()
+	if err != nil {
+		m.Errorf("init charts: %v", err)
+		return false
+	}
+	return true
+}
+
+func (m *Mongo) Check() bool {
+	return len(m.Collect()) > 0
+}
+
+func (m *Mongo) Charts() *module.Charts {
+	return m.charts
+}
+
+func (m *Mongo) Collect() map[string]int64 {
+	if m.client == nil {
+		var err error
+		err = m.initMongoClient()
+		if err != nil {
+			m.Errorf("init mongo client: %v", err)
+			return nil
+		}
+	}
+
+	ms := make(map[string]int64)
+	m.serverStatusCollect(ms)
+	if len(ms) == 0 {
+		m.Warning("zero collected values")
+		return nil
+	}
+	return ms
+}
+
+func (m *Mongo) Cleanup() {
+	if m.client == nil {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.Timeout*time.Second)
+	defer cancel()
+	err := m.client.Disconnect(ctx)
+	if err != nil {
+		m.Warningf("cleanup: error on closing mongo client: %v", err)
+	}
+	m.client = nil
+}
+
+func (m *Mongo) initMongoClient() error {
+	client, err := mongo.NewClient(options.Client().ApplyURI(m.Uri))
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.Timeout*time.Second)
+	defer cancel()
+	err = client.Connect(ctx)
+	if err != nil {
+		return err
+	}
+	m.client = client
+	return nil
```

Comment:
Do we need to call Ping here?

From [the documentation](https://github.com/mongodb/mongo-go-driver#usage)

> Calling Connect does not block for server discovery. If you wish to know if a MongoDB server has been found and connected to, use the Ping method

Reply:
I am not 100% sure about the ping. It does send a command to the database and adds a round trip. I was actually thinking replacing the `collect` with `ping` in the check method. If we do just the connect as we do it now, check will fail anyway because it calls collect if the database is down or the connection uri is wrong. What do you think?

Reply:
That is a good idea. One possible issue I can think of - ping (check) success, but collect fails, because of permissions.

Reply:
We consider a job active the job's Check returns `true` and keep retrying (call Collect) it indefinitely (but there is a retry backoff algorithm i believe). So in that case we get "check ok" in the logs, but no charts on the dashboard.

Reply:
Ok, it makes sense to leave the collect as it is then. Regarding the ping I think we don't need it.

Reply:
I agree! Not sure why I brought up this question 😅